src/lib/Microsoft.Health.Events/EventHubProcessor/PartitionLockingBackgroundService.cs

```diff
@@ -0,0 +1,429 @@
+
+                    throw;
+                }
+            }
+        }
+
+        private static Pageable<BlobItem> ListBlobs(BlobContainerClient blobContainerClient, string prefix)
+        {
+            try
+            {
+                var resultSegment = blobContainerClient.GetBlobs(BlobTraits.Metadata, BlobStates.None, prefix);
+                return resultSegment;
+            }
+            catch
+            {
+                throw;
+            }
+        }
+
+        private async Task<int> GetActiveProcessors(Pageable<BlobItem> resultSegment, BlobContainerClient containerClient, CancellationToken ct)
+        {
+            var activePods = new HashSet<string>();
+
+            foreach (BlobItem blobItem in resultSegment)
+            {
+                // if the processor has been updated in the last minute then consider it active
+                // otherwise delete it
+                if (blobItem.Metadata.TryGetValue("activeuntil", out var activeUntil))
+                {
+                    if (DateTime.Parse(activeUntil) > DateTime.UtcNow.AddMinutes(-1))
+                    {
+                        _logger.LogTrace($"Blob name: {blobItem.Name} is active");
+                        activePods.Add(blobItem.Name);
+                    }
+                    else
+                    {
+                        // todo: check the etag as a condition for delete
+                        _logger.LogTrace($"Deleting blob {blobItem.Name}. It was last active {activeUntil}.");
+                        await containerClient.DeleteBlobAsync(blobItem.Name);
+                    }
+                }
+            }
+
+            return activePods.Count;
+        }
+
+        public async Task RunAndRenew(CancellationTokenSource cts)
+        {
+            // in the background we will continually update partition ownership until
+            // we receive a cancellation token to shutdown the processor
+            _logger.LogTrace($"Running the processor with partitions: {string.Join(", ", _ownedPartitions.ToArray())}");
+
+            try
+            {
+                await RenewOwnership(cts.Token);
+                await CheckProcessorCount(cts);
+                await StartAssignedPartitionProcessor(cts.Token, _ownedPartitions);
+            }
+            catch (TaskCanceledException)
+            {
+                _logger.LogTrace("A task was cancelled");
+            }
+        }
+
+        protected override async Task ExecuteAsync(CancellationToken ct)
+        {
+            CreateContainerClient();
+            await RegisterProcessor();
+            await ContinuallyRegisterProcessor();
+
+            while (true)
+            {
+                using var innerCts = CancellationTokenSource.CreateLinkedTokenSource(ct);
+                while (!innerCts.IsCancellationRequested)
+                {
+                    await Task.Delay(5000);
+                    _eventHubPartitions = await GetPartitionsIdsForEventHub();
```

Comment:
we should also restart if this value changes. I believe there are some tiers of event hub ([premium, dedicated](https://learn.microsoft.com/en-us/azure/event-hubs/dynamically-add-partitions)) where the number of partitions can be changed.